\documentclass[11pt,a4paper,oneside,notitlepage]{book}
\usepackage{amsmath}                    % Uitgebreide wiskundige mogelijkheden
\usepackage{amssymb}                    % Voor speciale symbolen zoals de verzameling Z, R...
\usepackage{makeidx}                    % Om een index te maken
\usepackage{url}						% Om url's te verwerken
\usepackage{fixltx2e}
\usepackage{appendix}
\usepackage{textcomp}            
\usepackage{epstopdf}
\usepackage{graphicx}                   % Om figuren te kunnen verwerken
\usepackage[pdftex]{graphicx}
%\usepackage[small,bf,hang]{caption2}    % Om de captions wat te verbeteren
\usepackage{xspace}                     % Magische spaties na een commando
%\usepackage[latin1]{inputenc}           % Om niet ascii karakters rechtstreeks te
\usepackage{hyperref}
\usepackage{float}                      % Om nieuwe float environments aan te maken. Ook optie H!
%\usepackage{flafter}                   % Opdat floats niet zouden voorsteken
\usepackage{listings}                   % Voor het weergeven van letterlijke text en codelistings
\usepackage[round]{natbib}              % Voor auteur-jaar citaties.
\usepackage[nottoc]{tocbibind}			% Bibliografie en inhoudsopgave in ToC; zie tocbibind.dvi
\usepackage{marvosym}                   % Om het euro symbool te krijgen
%\usepackage{textcomp}                  % Voor onder andere graden celsius
%\usepackage{fancyhdr}                  % Voor fancy headers en footers.
%\usepackage[small,bf,hang]{caption2}	% Voor de captions te verbeteren
\usepackage[Gray,squaren,thinqspace,thinspace]{SIunits} % Elegant eenheden zetten 
\renewcommand{\baselinestretch}{1.35}   % De interlinie afstand wat vergroten.
%\usepackage[english,dutch]{babel}
\usepackage{flafter} %nieuw
\usepackage{lscape} %nieuw
\usepackage{fancyhdr}
\usepackage{graphicx}
\usepackage{amsmath, amssymb, amsbsy}
\usepackage{verbatim} 
\usepackage{float}
%\usepackage{subfigure}
\usepackage{subfig}
%\usepackage[small,bf,hang]{caption2}

% marges aanpassen
% (opmerking: moet *voor* inclusie van fancyhdr package komen)
\setlength{\hoffset}{-1in}
\setlength{\voffset}{-1in}
\setlength{\topmargin}{2cm}
\setlength{\headheight}{0.5cm}
\setlength{\headsep}{1cm}
\setlength{\oddsidemargin}{3.5cm}
\setlength{\evensidemargin}{3.5cm}
\setlength{\textwidth}{16cm}
\setlength{\textheight}{23.3cm}
\setlength{\footskip}{1.5cm}
\setlength{\parindent}{0cm}       
\setcounter{secnumdepth}{3} %nieuw


%Deftige fancy-header (hoofdstukken)
\pagestyle{fancy}
\fancyhf{}
\renewcommand{\headrulewidth}{0.5pt}
\fancyhf[HL]{\uppercase{\textit{\leftmark}}}
\fancyhf[HR]{\thepage}

\graphicspath{{images/}} % place of the figures

%\makeindex                              % Om een index te genereren.

\setcounter{MaxMatrixCols}{20}          % Max 20 kolommen in een matrix
\setcounter{tocdepth}{3}

% De headers die verschijnen bovenaan de bladzijden, herdefinieren:
\pagestyle{fancy}                       % Om aan te geven welke bladzijde stijl we gebruiken.
\fancyhf{}                              % Resetten van al de fancy settings.
\renewcommand{\headrulewidth}{0.4pt}      % Geen lijn onder de header. Zet dit op 0.4pt voor een mooie lijn.
\fancyhf[HL]{\nouppercase{\textit{\leftmark}}} % Links in de header zetten we de leftmark,
\fancyhead[HR]{\thepage}                % Rechts in de header het paginanummer.
% Activeer de volgende lijn en desactiveer de vorige om paginanummers onderaan gecentreerd te krijgen.
%\fancyhf[FC]{\thepage}                  % Paginanummers onderaan gecentreerd.

\newcommand{\cel}{$^{\circ}$C}


%%%%%%%%%%%%%%%%%%%%%%%%%
% Einde van de preamble.%
% Begin van de body:	  %
%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{document}


\frontmatter

\title{Protein Processing Resource manual}
\author{Niklaas Colaert}


%  Titelblad

\begin{titlepage}

\fontsize{12pt}{14pt}\selectfont

\begin{center}

\begin{figure}[H] 
		\begin{center}
			\includegraphics[width=3cm, height=4cm]{VIB.pdf}
			\hfill 
			\includegraphics{compomics.png}
			\hfill 
			\includegraphics{ugent.pdf}
		\end{center}
\end{figure}



\vspace{0.3cm}


\vspace{0.8cm}
Department of Medical Protein Research, VIB, B-9000 Ghent, Belgium
\linebreak Department of Biochemistry, Ghent University, B-9000 Ghent, Belgium
\linebreak  \url{http://www.proteomics.be}
\linebreak  \url{http://www.compomics.com}


\vspace{1.2cm}
\begin{figure}[H] 
		\begin{center}
			\includegraphics[]{ppr.jpg}
		\end{center}
\end{figure}

	\begin{minipage}[c][4.5cm][c]{15cm}
	\begin{center}
	\fontsize{17.28pt}{21pt}\selectfont
	{\textsc{TOPPR manual}}
	\fontseries{m}
	\vspace{0.2cm}
	\fontsize{12pt}{14pt}\selectfont
	\linebreak Niklaas Colaert
	\end{center}
	\end{minipage}
	

\vspace{3.5cm}

\date{}
 \url{http://iomics.ugent.be/toppr/}
\end{center}
\end{titlepage}



\tableofcontents

\mainmatter

\chapter*{Introduction}
\textbf{The Online Protein Processing Resource (\begin{large}
TOPPR
\end{large})}
\paragraph{}Over two percent of all human and mouse genes encode for proteases. This large group of enzymes controls many biological processes and they are crucially important in relatively simple processes such as food digestion as well as in highly regulated processes such as controlled cell death. In addition, proteases add to several pathologies including cancer, cardiovascular and inflammatory diseases. It is commonly recognized that a more detailed understanding of protease-controlled processes can be achieved by extending our overall knowledge on proteases, their (preferred) substrates and specificities. 
\paragraph{}The N-terminal COFRADIC (COmbinded FRActional DIagonal Chromatography) technique developed in our lab enables isolation and identification of protein N-terminal peptides using peptide chromatography and mass spectrometry. Given that protein processing induces new N-terminal protein ends, the resulting neo-N-terminal peptides are also isolated and identified and represent clear markers for the actual cleavage position in the protease substrate . Recently, a similar C-terminal COFRADIC technique was developed to select and identify neo C-terminal peptides, also identifying the processing event. These COFRADIC techniques made it possible to identify large numbers of processing events for several individual proteases or in cellular setups.
\paragraph{}Managing, analyzing, comparing and integrating processed events from different proteases have always been very cumbersome. Here we present The Online Protein Processing Resource (TOPPR) that stores high quality processed events and are made available in and easy and intuitive analysis platform.
\paragraph{}This TOPPR  user manual is divided in three parts. The first part (Searching TOPPR, chapter \ref{searching}) describes different search strategies that can be followed in the process of retrieving relevant data from TOPPR. The second part (Viewing TOPPR, chapter \ref{viewing}) illustrates and explains how TOPPR visualizes the search results and which type of analysis can be performed on the results. The last part (Additional terms, chapter \ref{common}) explains some common and useful terms.

%chapter on the searching TOPPR
\chapter{Searching TOPPR}\label{searching}

TOPPR is a complete protease substrate database. Three different search methods (parameter search (\ref{parameter}), UniProtKB/Swiss-Prot search (\ref{uniprot}) and motif search(\ref{motif})) were created and are briefly discussed in the following sections.
\section{Parameter search} \label{parameter}
\paragraph{}The parameter search is present to find large lists of substrates rather than presenting different processed sites in one protein. The simplest way of searching with the parameter search is selecting a treatment from the treatment window (see figure \ref{fig:Treatments}). Additional information concerning these treatments can be obtained by clicking the treatment(s) link. All the substrates of this treatment will be given in the result section of the search. Only  processed sites generated by the selected treatment(s) will be shown. processed sites found in the substrates that were not generated by the selected treatment(s) will not be shown.

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.7]{treatment.png}
	\caption{One or more treatments can be selected to perform a simple parameter search.}
	\label{fig:Treatments}
	\end{center}
\end{figure}

\paragraph{}Multiple treatments can be selected by holding down the control key and clicking the different treatments. Then, only proteins cleaved by all  selected treatments will be given in the search result section.
\paragraph{}Advanced search queries can be created with the parameter search. By clicking ``advanced search?" in the right bottom of the parameter search field, the advanced search options will be visualized.
\paragraph{}The first parameter that can be changed is the search type. The search type is the method TOPPR uses to merge different results. The default setting is ``And". Only the substrate overlap of the selected parameters (e.g. multiple treatments) will be shown as a result. If the search method is set to ``Or", all the substrates that are found for any search parameter will be shown. When the search method is set to ``And not" a positive selection drop-down menu appears. Only the substrates that are found for the selected treatment and not the substrates found using the other search parameters will be shown. The different search types are indicated in figure \ref{fig:SearchType}.

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{searchtype.png}
	\caption{The different parameter search types that can be selected.}
	\label{fig:SearchType}
	\end{center}
\end{figure}

\paragraph{}The second parameter that can be changed is ``same site search?". The default setting is ``No" indicating that the different selected treatments are not required to cleave the substrate on the same position. If same site search is set to ``Yes", only proteins that are cleaved by different selected treatments on the same position will be shown (\ref{fig:SameSite}).

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{samesite.png}
	\caption{The effect of using the ``Same site search" parameter.}
	\label{fig:SameSite}
	\end{center}
\end{figure}

\paragraph{}If ``same site search" is set to ``Yes", changing the search type from ``And" to ``Or" will have no effect. Changing the search type from ``And" to ``And not" on the other hand will result in an empty search result.
\paragraph{}The third parameter that can be set is the ``pre" and ``post" motif. Only the processed sites that match the motif will be included in the search result. More information about the motif search can be found in section \ref{motif}.
\paragraph{}By changing and selecting the last parameters (experimental type, cell source, taxonomy and inhibitor) specific experimental conditions can be selected and queried. These parameters are influenced by the search type. If the search type is set to ``Or" and there is one treatment and one taxonomy selected, all the proteins stored in TOPPR for this taxonomy and all the proteins cleaved by this treatment will be shown.

\section{UniProtKB/Swiss-Prot search} \label{uniprot}
\paragraph{}Searching for a specific substrate and its corresponding processed sites can be done with the UniProtKB/Swiss-Prot search. Three different search methods were developed to search the different UniProtKB/Swiss-Prot protein characteristics.
\begin{itemize}
	\item \textbf{UniProtKB/Swiss-Prot accession search:} Search for a specific UniProtKB/Swiss-Prot accession number. A substrate and its corresponding processed sites will be found if there is a protein stored in TOPPR with this UniProtKB/Swiss-Prot accession number. If no substrate is found for this accession an orthologue search will be performed. A RefSeq NCBI accession will first be identified for this UniProtKB/Swiss-Prot accession number. This is done by PICR (Protein Identifier Cross-Reference Service) a tool developed at the EBI. The orthologues and paralogues for this RefSeq protein can be found by using the HomoloGene NCBI database. HomoloGene is a system for automated detection of homologues (orthologues and paralogues) among annotated genes of several completely sequenced eukaryotic genomes. The found RefSeq orthologues and paralogues will be ``translated" back by PICR to a UniProtKB/Swiss-Prot accession. This accession will finally be searched in TOPPR. The diagram (see figure \ref{fig:Ortho}) summarizes this orthologue finding method.
\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.9]{ortho.png}
	\caption{The orthologue finding method.}
	\label{fig:Ortho}
	\end{center}
\end{figure}
If an orthologue or paralogue is found in TOPPR the protein will be visualized in a red box, and not in a black box (see figure \ref{fig:Box}).
\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{box.png}
	\caption{The box at the top is the normal representation of a found substrate. The box at the bottom with the red halo is shown when an orthologue or paralogue is found.}
	\label{fig:Box}
	\end{center}
\end{figure}

	
	\item UniProtKB/Swiss-Prot entry name search: Search for a specific UniProtKB/Swiss-Prot entry name.\textit{ Example: search for ACTB\_HUMAN.}
	\item UniProtKB/Swiss-Prot description search: Searh for a term in the UniProtKB/Swiss-Prot name or description of a protein.\textit{ Example: search for \textbf{\begin{large}
actin\end{large}}. TOPPR will retrieve several substrates. Obviously it will find \textbf{\begin{large}
Actin
\end{large}}, cytoplasmic 1 - Homo sapiens and \textbf{\begin{large}
Actin
\end{large}} cytoplasmic 1 – Mus musculus but also \textbf{Actin}-related protein 2 - Homo sapiens and C-jun-amino-terminal kinase-inter\textbf{\begin{large}
actin\end{large}}g protein 4 - Homo sapiens.}
\end{itemize}

\section{Motif search} \label{motif}
\paragraph{}The amino acids surrounding a processed site are often very important for substrate recognition by a protease. Searching with a specific amino acid motif before (``pre-site motif") and after (``post-site motif") the processed site can be done in TOPPR. Different search strategies can be used. The easiest strategy is to search for a specific sequence of amino acids.\textit{ Example: Pre-site motif = DEVD. TOPPR will only report processed sites where the non-prime (P4-P1) amino acids are DEVD. Pre-site motif = DEVD, post-site motif = G. TOPPR will only find processed sites where the non-prime amino acids (P4-P1) are DEVD and the prime amino acid (P1') is G.} 
\paragraph{}Another method to find specific processed sites is to use the character *. Any amino acid can occur at the position of *. \textit{Example: Pre-site motif = D**D. TOPPR will find processed site where the non-prime amino acids are DEVD, DSED, DGAD,... .} 
\paragraph{}Searching for two specific amino acids on one position  can be done by searching for [AA 1/AA 2]. \textit{Example: pre-site motif = DE[V/I]D. TOPPR will only find processed site where the non-prime amino acids are DEVD or DEID.}


%chapter on the viewing OPPR
\chapter{Viewing TOPPR}\label{viewing}

\paragraph{}Four different sections (search summary (\ref{summaryView}), motif section (\ref{motifView}), peptide section (\ref{peptideView}) and protein section (\ref{proteinView})) appear in a search result. At the bottom of every search result page there is an user protein compilation bar (\ref{bar}). Statistics concerning TOPPR can be found  \htmladdnormallink{here}{http://iomics.ugent.be/toppr/Statistics} (this takes sometime to calculate!). 
 
\section{Search summary} \label{summaryView}
\paragraph{}Every search result window starts with a search summary. The number of substrates found and the number of corresponding processed sites are indicated. A peptide can be mapped on two different proteins. When this happens, TOPPR will store two different processed sites. The number of unique peptides found in a search result is counted and is given in the summary. The search parameters are shown and differ depending on the used search method. Links to the other parts of the search result are also present in the search summary.
 
\section{Motif section} \label{motifView}
\paragraph{}Search results obtained using the parameter or motif search method have a motif section. Also, proteins in the user compilation that are viewed have a motif section.
\paragraph{}In the motif section different tools can be used to analyse the found substrates.
\begin{itemize}
	\item \textbf{Automatic generation of an iceLogo.} An iceLogo (see \ref{iceLogo}) needs both a positive or experimental set, and a negative or background set. For both the experimental and background set, one or more treatments can be selected. The background set can also be set as the natural occurrence of amino acids from the substrate species.  A subset of these processed sites will be used to generate the iceLogo if the ``Only selected peptides?" check box is selected, and these peptides or processed sites can be selected in the peptide section. The default P value used in the generation of an iceLogo is 0.05. The sites in the processed site list can be used to create other iceLogo visualization on \url{http://iomics.ugent.be/icelogoserver/logo.html}.
	\item \textbf{Automatic generation of a weblogo.} One or more treatments can be selected. The processed sites that are present in the search result and that are cleaved by the selected treatments will be used in the generation of the weblogo (see \ref{weblogo}). A subset of these processed sites will be used to generate the weblogo if the ``Only selected peptides?" checkbox is selected. The peptides or processed site can be selected in the peptide section. The height of the Y-axis in the weblogo can be changed (default is 4.0 bits) by selecting a different number in the drop down menu ``Weblogo bits".
The width of the weblogo can be changed (default is from P15 to P15`) in the ``Weblogo width" drop down menu. A choice can be made between 15, 10, 8, 4 and 2 amino acids pre (non-prime sites) and post (prime sites) processed site.
\\
The unchangeable default settings for the creation of a weblogo are:
\begin{itemize}
	\item picture format = PNG
	\item weblogo width = 500 px
	\item weblogo height = 300 px
	\item small sample correction = on
	\item color scheme = default
	\item frequence plot = off
	\item show fineprint = on
\end{itemize}
	\item \textbf{Starting the Jalview applet.} One or more treatments can be selected. The processed sites that are present in the search result and that were cleaved by the selected treatments will be visualized with Jalview. A subset of these processed sites will be visualized if the ``Only selected peptides?" check box is selected, and peptides are selected.
	\item \textbf{Generating a POPS model.} One or more treatments can be selected. The processed sites that are present in the search result and that are cleaved by the selected treatments will be used in generating the Pops model (see \ref{popsTerm}). A subset of these processed sites will be used to generate the Pops model if the ``Only selected peptides?" checkbox is selected. The width of the Pops model can be changed (default from P8 to P8’) in the ``Pops model width" drop-down menu. For every position, all the amino acids will be counted. This number will then be divided by the total number of amino acids on that position. This number will be multiplicated by 10 after a subtraction of 0.5. An example of how a Pops model is calculated is shown in the following table.


\begin{table}[H] \label{pops}
\begin{center}
\begin{tabular}{|c|c|c|c|c|} 
\hline Amino acid & \# & \# /total & (\# /total)-0.5 & ((\# /total)-0.5)*10 \\ 
\hline A & 7 & 7/55 = 0.1272 & -0.3727 & -3.727 \\ 
\hline C & 1 & 1/55 = 0.0181 & -0.4818 & -4.818 \\ 
\hline D & 30 & 30/50 = 0.5454 & 0.0454 & 0.454 \\ 
\hline ... & 17 & ... & ... & ... \\ 
\hline Y & 0 & 0 & -0.5 & -5 \\
\hline 
\end{tabular}
\caption{The numbers in the last column will be used in the Pops model}
\end{center}
\end{table}	
\end{itemize}

Only distinct, non-redundant peptides (with the processed sites in the middle) will be used to generate iceLogos, weblogos and are viewable in the processed site list. This is done because otherwise peptides that are mapped to multiple proteins (isoforms) will get an unwanted higher weight and influence in the logos.
\textit{Example: The peptide SGFTSLLMERL is mapped to three different proteins (Tubulin alpha-1B chain - Homo sapiens, Tubulin alpha-1C chain - Homo sapiens and Tubulin alpha chain-like 3 - Homo sapiens). Only two peptides (with the processed sites in the middle) will be used to generate the weblogo. The first and the second processed peptide are the same (both LQGFLVFHSFGGGTGSGFTSLLMERLSVDY). Only one of those two will be selected. The third peptide is not equal to the first two (the amino acids that differ are indicated in bold). The last peptide will therefore also be used to generate the weblogo although it was mapped by the same peptide (SGFTSLLMERL).
\begin{footnotesize}
\linebreak Tubulin alpha-1B chain - Homo sapiens 131-LQGFLVFHSFGGGTG (146) \underline{SGFTSLLMER}LSVDY-161
\linebreak Tubulin alpha-1C chain - Homo sapiens 131-LQGFLVFHSFGGGTG (146) \underline{SGFTSLLMER}LSVDY-161
\linebreak Tubulin alpha chain-like 3 - Homo sapiens 138-LQGFL\textbf{I}F\textbf{R}SFGGGTG (153) \underline{SGFTSLLMER}L\textbf{TGEY}-168
\end{footnotesize}}
 


\section{Peptide section} \label{peptideView}
\paragraph{}A peptide section is generated by a parameter search, motif search and a user protein compilation. All the found processed sites are listed in the peptide section. These processed sites can be grouped by treatment or by protein. Ordering by protein can give more information about the number of processed sites in a protein. Processed sites cleaved by multiple treatments can be easily spotted in this type of listing. The number of processed sites found for a specific treatment can best be seen in the treatment ordering. In both ordering methods processed sites or peptides can be selected. These selected peptides can be used to generate a subset of processed sites. This subset can be analysed by different tools in the motif section. Substrates in the peptide section can then be added to a ``user protein compilation". This compilation can be viewed by opening the performed searches field.
 
\section{Protein section} \label{proteinView}
\paragraph{}All the search methods expect the parameter search method generate a protein section. For every substrate there is a box with three fields: the sequence, the processed sites and the extra information field. This box will be red if an orthologue or paralogue was found (See UniProtKB/Swiss-Prot search \ref{uniprot} for more information).
\subsection{The sequence field}
\paragraph{}At the top of the field the UniProtKB/Swiss-Prot accession and the UniProtKB/Swiss-Prot entry name are given. The next line shows the UniProtKB/Swiss-Prot description for this protein. At the bottom of this field a bar is created. This bar represents the primary protein structure. Blue and red squares indicate the relative position of the processed sites in this protein. The squares are blue when two or more treatments induced a processed event on that position. If only one treatment cleaves the protein there, the square is red. Hovering over the squares will generate visualization of text, which gives the processed site position and the treatment(s). Clicking the square will generate the visualization of the processed site in the protein sequence by a colored scissor. The scissors can be made invisible by clicking the square again. Figure \ref{sequence} gives an impression of the sequence field.

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{sequence.png}
	\caption{The sequence field.}
	\label{sequence}
	\end{center}
\end{figure}

\paragraph{}Next to the sequence two different drawers give additional functions. In the first ``Options" drawer, three different actions can be initiated.
\begin{itemize}
\item \textbf{Show domains} This action will visualize the know domains in the PFAM and SMART database by visualizing a domain protein bar under the unfolded protein picture. Hovering over the domains will indicate the name and position of the protein domain. Clicking it will open a new tab with the information about the domain in the PFAM or SMART database.
\item \textbf{Show secondary structure} A very simple secondary structure predication (\textit{via} \htmladdnormallink{StrBioLib}{http://bioinformatics.oxfordjournals.org/content/23/15/2018.full}) will be indicated on the sequences. Beta sheet predications are indicated by a green arrow whereas alpha helices are indicated by a blue loop.
\item \textbf{Show Gene Ontologies} The know gene ontology terms linked to this protein will be visualized beneath the sequence. Clicking on the link will open a new tab with the gene ontology information on the gene ontology web site.
\item \textbf{View the 3D structure in JMol} This action can only be done if (a part) of the structure of the protein is known. A different structure can be selected if more than one pdb structure is found by choosing one in the radio button list. By default the P1 position (the amino acid just before the cleavage site) is visualized in the structure by a green ball and stick representation of the corresponding amino acid. Other (non) primed positions can be selected below the structure radio button list (see figure \ref{jmol}). 

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{jmol.png}
	\caption{Example 3D visualization in TOPPR.}
	\label{jmol}
	\end{center}
\end{figure}

\item \textbf{View information about the spectra and peptides} Every processed site is identified by either the N- or C-terminal COFRADIC technology. This implies that every processed site is linked to an identified MS/MS spectrum. Information about the identifications and spectra can be found by pressing this link. The (B and Y ion annotated) spectrum can be visualized by making use of the Pride spectrum viewer.
\item \textbf{View the conservation of the processed sites} The orthologues (found by the method described in \ref{uniprot}) will be aligned by the Needleman-Wunsch pairwise global alignment algorithm with default parameters. An example output of this is given in figure \ref{conservation}.

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{conservation.png}
	\caption{Example output for the processed site conservation.}
	\label{conservation}
	\end{center}
\end{figure}
\end{itemize}
\linebreak
The second drawer ``Treatments" holds a list of buttons that can be used to show and hide the processed sites of specific treatments in the protein sequence.
\subsection{The processed site field}
\paragraph{}Every processed site and the peptide with the sequence surrounding this site is listed and ordered by treatment. The processed sites can also be ordered by processed position in the protein. Ordering by position can be used to find positions where more than one treatment cleaved the protein. The processed site sequence is in italics if different isoforms were found for the MS/MS identified peptide, mapped on the substrate. Hovering over such a peptide will generate a small box displaying the number of isoforms found. Clicking it will start a motif search for this peptide and thus generates a search result with all the isoforms. Figure \ref{procsites} gives an impression of the processed site field.

\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.5]{procsites.png}
	\caption{The processed sites field.}
	\label{procsites}
	\end{center}
\end{figure}

 
\section{User protein compilation bar} \label{bar}
\paragraph{}At the bottom of the result screen, a green bar is shown (see figure \ref{greenbar}). A table with the performed searches is visualized if ``View performed searches" is clicked. The result of these searches can be viewed in a new window or a new tab if the search parameters in the first column are clicked. In this green bar, the proteins that are in the user protein compilation are shown. Proteins can be added by clicking ``add" in the peptide section. Proteins can be deleted again from this compilation by clicking the white cross next to the protein UniProtKB/Swiss-Prot accession in the green bar.
The user protein compilation is a useful tool to compare different proteins of interest. The compilation can be viewed by going to the performed searches table and clicking on the compilation. Performed searches can also be visualized by clicking the search log button in the upper bar.


\begin{figure}[H]
	\begin{center}
	\includegraphics[scale=0.7]{greenbar.png}
	\caption{The user compilation bar.}
	\label{greenbar}
	\end{center}
\end{figure}

%chapter on common terms
\chapter{Additional terms}\label{common}

\section{iceLogo} \label{iceLogo}
\paragraph{}IceLogo builds on probability theory to visualize significant conserved sequence patterns in multiple peptide sequence alignments against background (reference) sequence sets that can be tailored to the studied system and the used protocol. The unique advantages of iceLogo compared to other sequence logo creating tools is that iceLogo has a more dynamic nature and is correcter and completer in the analysis of conserved sequence patterns. Click \htmladdnormallink{here}{http://iomics.ugent.be/icelogoserver/} for more information.

\linebreak
\paragraph{}\textit{Improved visualization of protein consensus sequences by iceLogo. Colaert N, Helsens K, Martens L, Vandekerckhove J, Gevaert K. Nat Methods. 2009 Nov;6(11):786-7}

\section{Weblogo} \label{weblogo}
\paragraph{}Weblogo is a web-ased application designed to make the generation of sequence logos as easy and painless as possible. Sequence logos are a graphical representation of an amino multiple sequence alignment developed by Tom Schneider and Mike Stephens. Each logo consists of stacks of symbols, one stack for each position in the sequence. The overall height of the stack indicates the sequence conservation at that position, while the height of symbols within the stack indicates the relative frequency of each amino or nucleic acid at that position. In general, a sequence logo provides a richer and more precise description of a processed site than would a consensus sequence. Click \htmladdnormallink{here}{http://weblogo.berkeley.edu/} for more information.

\linebreak
\paragraph{}\textit{Sequence logos: a new way to display consensus sequences.Schneider TD, Stephens RM. Nucleic Acids Res. 1990 Oct 25;18(20):6097-100.}
\paragraph{}\textit{WebLogo: a sequence logo generator. Crooks GE, Hon G, Chandonia JM, Brenner SE. Genome Res. 2004 Jun;14(6):1188-90.}


\section{PoPS} \label{popsTerm}
\paragraph{}PoPS is a software tool for modeling protease activity. Its main aim is to provide an environment in which the user can learn and reason about the activity of any protease. PoPS allows the user to create a model of activity for an arbitrary protease by specifying general parameters that are commonly used to describe the function of a protease, for example, the number of important subsites, their amino acid preferences and the relative importance of the subsites. The user can then supply a substrate (protein sequence) and PoPS will use the protease model to predict where the substrate will be cleaved.
A simple prediction of cleavages is very useful in itself, particularly if the protease has a well-defined activity. In addition, the predictions do not just include what will happen, but also what will not happen. However, PoPS's utility does not end here. Once a prediction has been made, PoPS allows the user to adjust parameters and the original protease model to perform a deeper investigation of how the protease is functioning. For example, in the case of a protease with a poorly defined function, predictions can be compared to experimental results, and the model can be altered to gain a better understanding of the true model of the protease's activity.
\paragraph{}PoPS also aims to provide other information about the substrate that might improve the accuracy of predictions. This includes tertiary structure and secondary structure predictions, but more options are currently being implemented. Click \htmladdnormallink{here}{http://pops.csse.monash.edu.au/pops-cgi/index.php} for more information.
\linebreak
\paragraph{}\textit{PoPS: a computational tool for modeling and predicting protease specificity. Boyd SE, Pike RN, Rudy GB, Whisstock JC, Garcia de la Banda M. J Bioinform Comput Biol. 2005 Jun;3(3):551-85.}


\backmatter


\end{document}